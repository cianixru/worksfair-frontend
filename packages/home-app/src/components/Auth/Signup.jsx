import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import { Helmet } from 'react-helmet';

import SignupForm from '../../forms/Auth/SignupForm';
import {
  setAuthenticatedUser,
  AUTHENTICATION_FAILED,
  AUTHENTICATED_USER,
} from '../../actions/auth';
import { isLoading, isComplete } from '../../actions/loader';
import alert from '../utils/alert';
import { baseURL } from '../../utils/api';
import { sendConfirmationEmail } from '../../utils/services';

class Signup extends Component {
  state = {
    validationErrors: {
      first_name: [],
      last_name: [],
      email: [],
      username: [],
    },
  }

  /**
   * @description handles the reset of the error detail
   * @param { string } name
   */
  handleErrorReset = (name) => {
    this.setState({
      validationErrors: {
        ...this.state.validationErrors,
        [name]: [],
      },
    });
  };

  onSubmit = async (details) => {
    const { actions, history } = this.props;
    try {
      await actions.isLoading();

      const url = `${baseURL}/auth/register/`;
      details.email = details.email.toLowerCase();
      details.role = 'admin';
      let type = AUTHENTICATED_USER;
      fetch(url, {
        method: 'POST',
        body: JSON.stringify(details),
        headers:{
          'Content-Type': 'application/json'
        }
      }).then(res => {
        if (!res.ok) {
          type = AUTHENTICATION_FAILED;
          alert.error('Signup failed. Check for more details');
        }
        return res.json();
      })
      .then(response => {
        if (type === AUTHENTICATION_FAILED) {
          const { user } = response;
          this.setState({
            validationErrors: user,
          });
          if (user.non_field_errors) {
            alert.info(user.non_field_errors[0]);
          }
        } else {
          alert.success('Successfully Signed up!');

          const { user: { email, token, username } } = response;
          const url = `${
            window.location.host}/dashboard/${
            username}/webpages?token=${token}`;

          if (token && email) {
            sendConfirmationEmail(url, email);
            history.push({
              pathname: `/dashboard/${username}/webpages`,
              state: { previousLocation: 'signup' }
            });
          } else {
            alert.warning('Something went wrong. Confirmation email was not sent.');
            history.push({
              pathname: `/dashboard/${username}/webpages`,
              state: { previousLocation: 'login' }
            });
          }
        }
        const payload = {
          type,
          data: response,
        };
        actions.setAuthenticatedUser(payload);
      })
      .catch(error => console.error('Error:', error));

    } catch (error) {
      console.log(error);
    } finally {
      await actions.isComplete();
    }
  };

  render() {
    const { validationErrors } = this.state;
    return (
      <div className="auth-content signup">
        <Helmet>
          <title>Sign up - Worksfair</title>
        </Helmet>
        <h3 className="subtitle is-3">Sign up</h3>
        <SignupForm
          onSubmit={this.onSubmit}
          validationErrors={validationErrors}
          handleErrorReset={this.handleErrorReset}
        />
      </div>
    );
  }
}

Signup.propTypes = {
  actions: PropTypes.object,
  history: PropTypes.object,
  user: PropTypes.object,
  location: PropTypes.object,
};

const mapStateToProps = ({ auth: { currentUser } }) => ({
  user: currentUser.user,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    {
      setAuthenticatedUser,
      isLoading,
      isComplete,
    },
    dispatch,
  ),
});

export default withRouter(connect(
  mapStateToProps,
  mapDispatchToProps,
)(Signup));
