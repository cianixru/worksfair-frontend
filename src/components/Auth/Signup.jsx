import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';

import SignupForm from '../../forms/Auth/SignupForm';
import { signup } from '../../actions/auth';

class Signup extends Component {
  onSubmit = async (user) => {
    const { actions, history } = this.props;
    try {
      await actions.signup(user);
      history.push('/dashboard');
    } catch (err) {
      console.log(err);
    }
  };

  render() {
    return (
      <div>
        <h3 className="subtitle is-3">Sign up</h3>
        <SignupForm
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

Signup.propTypes = {
  actions: PropTypes.object,
  history: PropTypes.object,
};

export const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    {
      signup,
    },
    dispatch,
  ),
});

export default connect(
  null,
  mapDispatchToProps,
)(withRouter(Signup));
