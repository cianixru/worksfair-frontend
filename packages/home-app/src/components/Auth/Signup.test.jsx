import React from 'react';
import { Provider } from 'react-redux';
import { render, cleanup, fireEvent } from 'react-testing-library';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { BrowserRouter as Router } from 'react-router-dom';

import Signup from './Signup';

afterEach(cleanup);

const actions = {
  signup: jest.fn(),
};
const middlewares = [thunk];
const store = configureMockStore(middlewares);
const mockStore = store({ auth: { currentUser: {} } });

describe('Signup.jsx', () => {
  test('renders correctly', () => {
    const component = render(
      <Provider store={mockStore}>
        <Router>
          <Signup actions={actions} />
        </Router>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  test('should register a user', async () => {
    const { getByText, getByTestId } = render(
      <Provider store={mockStore}>
        <Router>
          <Signup actions={actions} />
        </Router>
      </Provider>
    );

    const firstName = getByTestId('firstName');
    const lastName = getByTestId('lastName');
    const email = getByTestId('email');
    const password = getByTestId('password');

    fireEvent.change(firstName, { target: { value: 'Theo' } });
    expect(firstName.value).toEqual('Theo');

    fireEvent.change(lastName, { target: { value: 'Okafor' } });
    expect(lastName.value).toEqual('Okafor');

    fireEvent.change(email, { target: { value: 'theo.okafor@worksfair.com' } });
    expect(email.value).toEqual('theo.okafor@worksfair.com');

    fireEvent.focus(email);
    expect(email.value).toEqual('theo.okafor@worksfair.com');

    fireEvent.change(password, { target: { value: 'password1$' } });
    expect(password.value).toEqual('password1$');

    const submitBtn = getByText('Signup');
    await fireEvent.click(submitBtn);
  });

  test.skip('should map dispatch to prop correctly', () => {
    const dispatch = () => { };
    const componentState = mapDispatchToProps(dispatch);
    expect(componentState.actions.signup).toBeDefined();
  });
});
