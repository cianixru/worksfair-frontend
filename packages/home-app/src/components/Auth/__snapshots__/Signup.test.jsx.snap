// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup.jsx renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h3
          class="subtitle is-3"
        >
          Sign up
        </h3>
        <form>
          <div
            class="field"
          >
            <label>
              First Name
            </label>
            <div
              class="control has-icons-left"
            >
              <input
                class="input is-medium"
                data-testid="firstName"
                name="first_name"
                placeholder="First Name"
                type="text"
                value=""
              />
              <span
                class="icon is-medium is-left"
              >
                <i
                  class="fa fa-user-o"
                />
              </span>
            </div>
          </div>
          <div
            class="field"
          >
            <label>
              Last Name
            </label>
            <div
              class="control has-icons-left"
            >
              <input
                class="input is-medium"
                data-testid="lastName"
                name="last_name"
                placeholder="Last Name (Surname)"
                type="text"
                value=""
              />
              <span
                class="icon is-medium is-left"
              >
                <i
                  class="fa fa-user-o"
                />
              </span>
            </div>
          </div>
          <div
            class="field"
          >
            <label>
              Username
            </label>
            <div
              class="control has-icons-left"
            >
              <input
                class="input is-medium"
                data-testid="username"
                name="username"
                placeholder="Username"
                type="text"
                value=""
              />
              <span
                class="icon is-medium is-left"
              >
                <i
                  class="fa fa-user-o"
                />
              </span>
            </div>
          </div>
          <div
            class="field"
          >
            <label>
              Email
            </label>
            <div
              class="control has-icons-left"
            >
              <input
                class="input is-medium"
                data-testid="email"
                name="email"
                placeholder="Email"
                type="text"
                value=""
              />
              <span
                class="icon is-medium is-left"
              >
                <i
                  class="fa fa-envelope-o"
                />
              </span>
            </div>
          </div>
          <div
            class="field"
          >
            <label>
              Password
            </label>
            <div
              class="control has-icons-left"
            >
              <input
                class="input is-medium"
                data-testid="password"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
              <span
                class="icon is-medium is-left"
              >
                <i
                  class="fa fa-lock"
                />
              </span>
            </div>
          </div>
          <div
            class="field"
          >
            Already signed up? 
            <a
              href="/login"
            >
              Log in
            </a>
          </div>
          <div
            class="control"
          >
            <button
              class="button is-link is-medium is-rounded"
              disabled=""
              type="submit"
            >
              Signup
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h3
        class="subtitle is-3"
      >
        Sign up
      </h3>
      <form>
        <div
          class="field"
        >
          <label>
            First Name
          </label>
          <div
            class="control has-icons-left"
          >
            <input
              class="input is-medium"
              data-testid="firstName"
              name="first_name"
              placeholder="First Name"
              type="text"
              value=""
            />
            <span
              class="icon is-medium is-left"
            >
              <i
                class="fa fa-user-o"
              />
            </span>
          </div>
        </div>
        <div
          class="field"
        >
          <label>
            Last Name
          </label>
          <div
            class="control has-icons-left"
          >
            <input
              class="input is-medium"
              data-testid="lastName"
              name="last_name"
              placeholder="Last Name (Surname)"
              type="text"
              value=""
            />
            <span
              class="icon is-medium is-left"
            >
              <i
                class="fa fa-user-o"
              />
            </span>
          </div>
        </div>
        <div
          class="field"
        >
          <label>
            Username
          </label>
          <div
            class="control has-icons-left"
          >
            <input
              class="input is-medium"
              data-testid="username"
              name="username"
              placeholder="Username"
              type="text"
              value=""
            />
            <span
              class="icon is-medium is-left"
            >
              <i
                class="fa fa-user-o"
              />
            </span>
          </div>
        </div>
        <div
          class="field"
        >
          <label>
            Email
          </label>
          <div
            class="control has-icons-left"
          >
            <input
              class="input is-medium"
              data-testid="email"
              name="email"
              placeholder="Email"
              type="text"
              value=""
            />
            <span
              class="icon is-medium is-left"
            >
              <i
                class="fa fa-envelope-o"
              />
            </span>
          </div>
        </div>
        <div
          class="field"
        >
          <label>
            Password
          </label>
          <div
            class="control has-icons-left"
          >
            <input
              class="input is-medium"
              data-testid="password"
              name="password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
            <span
              class="icon is-medium is-left"
            >
              <i
                class="fa fa-lock"
              />
            </span>
          </div>
        </div>
        <div
          class="field"
        >
          Already signed up? 
          <a
            href="/login"
          >
            Log in
          </a>
        </div>
        <div
          class="control"
        >
          <button
            class="button is-link is-medium is-rounded"
            disabled=""
            type="submit"
          >
            Signup
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
